src/analyzer/protocol/quic/QUIC.spicy

```diff
@@ -22,19 +22,81 @@ public function decrypt_crypto_payload(
 ##############
 
 # Can we decrypt?
-function can_decrypt(long_header: LongHeaderPacket, context: ConnectionIDInfo, is_client: bool): bool {
+function can_decrypt(long_header: LongHeaderPacket, context: ConnectionIDInfo&, state: CryptoState&): bool {
 
   if ( ! long_header.is_initial )
     return False;
 
-  if ( is_client )
-    return ! context.client_initial_processed;
+  # Can only decrypt the responder if we've seen the initial destination conn id.
+  if ( ! state.is_orig && |context.initial_destination_conn_id| == 0 )
+    return False;
+
+  # Only attempt decryption if we haven't flushed some SSL data yet.
+  return ! state.flushed;
+}
+
+# Buffer all CRYPTO frames, which might be fragmented and unordered
+# across multiple INITIAL packets into this CyptoBuffer unit.
+type CryptoBuffer = unit {
+  var buffered: bytes;
+  var length: optional<uint32>;
+  var complete: bool = False;
+
+  : bytes &chunked &eod {
+    self.buffered += $$;
+
+    # The first 4 bytes of crypto data contain the expected tag and a
+    # 24bit length from the TLS HandshakeMessage. Extract the length
+    # so we can determine when all CRYPTO frames have arrived.
+    #
+    # https://datatracker.ietf.org/doc/html/rfc8446#section-4
+    #
+    #  struct {
+    #    HandshakeType msg_type;    /* handshake type */
+    #    uint24 length;             /* remaining bytes in message */
+    #    ...
+    #
+    if ( ! self.length && |self.buffered| >= 4 ) {
+      # XXX: This looks terrible
+      self.length = (cast<uint32>(*self.buffered.at(1)) << 16) + (cast<uint32>(*self.buffered.at(2)) << 8) + (cast<uint32>(*self.buffered.at(3)));
+    }
+
+    if ( self.length && |self.buffered| >= uint64(*self.length) + 4 ) {
+      self.complete = True;
+    }
+  }
+};
+
+type CryptoState = struct {
+  is_orig: bool &default=False;
+  flushed: bool &default=False;
+  buf: CryptoBuffer&;
+  snk: sink&;
+};
+
+function crypto_state_complete(state: CryptoState&): bool {
+  return state.buf != Null && state.buf.complete;
+}
+
+function crypto_state_flush(context: ConnectionIDInfo&, state: CryptoState&) {
```

Comment:
Any reason we need to pass `context` as explicit ref? By default this should just work. Also, instead of relying on implicit interior mutability for refs for modifying `state` we should be able to modify it by explicitly passing it as `inout`.

```suggestion
function crypto_state_flush(context: ConnectionIDInfo, inout state: CryptoState) {
```

Reply:
> Any reason we need to pass context as explicit ref?

I'm not Spicy-versed enough to understand what the difference is without looking at the generated C++ code. That is not my workflow - it could, but I'm playing naive. I'm usually stopping at what works and what I might have used elsewhere (`CryptoState&` within Context). If one should be strictly preferred over the other and/or the `&` is redundant, compiling the analyzer probably should result in a warning/hint that that's so.

Reply:
The function is gone. I tried to remove `&` elsewhere, but it seems `inout` only works with units, not with structs, producing the following:
```
[error] /home/awelzel/corelight-oss/zeek/src/analyzer/protocol/quic/QUIC.spicy:476:39-476:69: unsupported type for unit parameter 'context': type of inout unit parameters must itself be a unit; for other parameter types, use references instead of inout
```

(that was for another unit rather than a function).